src/GitHub.InlineReviews/Glyph/GlyphMargin.cs

```diff
@@ -0,0 +1,215 @@
+                foreach (var span in e.Spans)
+                {
+                    list.AddRange(span.GetSpans(textView.TextSnapshot));
+                }
+
+                if (list.Count > 0)
+                {
+                    var span = list[0];
+                    int start = span.Start;
+                    span = list[0];
+                    int end = span.End;
+                    for (int i = 1; i < list.Count; i++)
+                    {
+                        span = list[i];
+                        start = Math.Min(start, span.Start);
+                        span = list[i];
+                        end = Math.Max(end, span.End);
+                    }
+
+                    var rangeSpan = new SnapshotSpan(textView.TextSnapshot, start, end - start);
+                    visualManager.RemoveGlyphsByVisualSpan(rangeSpan);
+                    foreach (var line in textView.TextViewLines.GetTextViewLinesIntersectingSpan(rangeSpan))
+                    {
+                        RefreshGlyphsOver(line);
+                    }
+                }
+            }
+        }
+
+        void OnLayoutChanged(object sender, TextViewLayoutChangedEventArgs e)
+        {
+            RefreshMarginVisibility();
+
+            visualManager.SetSnapshotAndUpdate(textView.TextSnapshot, e.NewOrReformattedLines, e.VerticalTranslation ? (IList<ITextViewLine>)textView.TextViewLines : e.TranslatedLines);
+
+            var lines = refreshAllGlyphs ? (IList<ITextViewLine>)textView.TextViewLines : e.NewOrReformattedLines;
+            foreach (var line in lines)
+            {
+                visualManager.RemoveGlyphsByVisualSpan(line.Extent);
+                RefreshGlyphsOver(line);
+            }
+
+            refreshAllGlyphs = false;
+        }
+
+        void OnZoomLevelChanged(object sender, ZoomLevelChangedEventArgs e)
+        {
+            refreshAllGlyphs = true;
+            marginVisual.LayoutTransform = e.ZoomTransform;
+        }
+
+        void RefreshGlyphsOver(ITextViewLine textViewLine)
+        {
+            foreach (IMappingTagSpan<TGlyphTag> span in tagAggregator.GetTags(textViewLine.ExtentAsMappingSpan))
+            {
+                NormalizedSnapshotSpanCollection spans;
+                if (span.Span.Start.GetPoint(textView.VisualSnapshot.TextBuffer, PositionAffinity.Predecessor).HasValue &&
+                    ((spans = span.Span.GetSpans(textView.TextSnapshot)).Count > 0))
+                {
+                    visualManager.AddGlyph(span.Tag, spans[0]);
+                }
+            }
+        }
+
+        void RefreshMarginVisibility()
+        {
+            marginVisual.Visibility = isMarginVisible(textView.TextBuffer) ? Visibility.Visible : Visibility.Collapsed;
+        }
+
+        void ThrowIfDisposed()
+        {
+            if (isDisposed)
+            {
+                throw new ObjectDisposedException(marginName);
+            }
+        }
+
+        public bool Enabled
```

Comment:
This might be personal preference, and I know we don't stick to it elsewhere, but personally I like methods to grouped together by access visiblility - when I looking through the code I assume the public API has finished when I get to private methods...